File: leetcode/76.最小覆盖子串/76.py
class Solution:
    def __init__(self):
        self.ori = defaultdict(int)
        self.cnt = defaultdict(int)

    def check(self)->bool:
        for key, val in self.ori.items():
            if self.cnt[key] < val:
                return False
        return True

    def minWindow(self, s: str, t: str) -> str:
        for i in range(len(t)):
            self.ori[t[i]] += 1
        left, right, sub, i = 0, 0, 0, 0
        res = sys.maxsize
        for j in range(len(s)):
            self.cnt[s[j]] += 1
            while self.check():
                sub = j - i + 1
                if sub < res:
                    left = i
                    right = j
                    res = sub
                self.cnt[s[i]] -= 1
                i += 1
        return "" if res == sys.maxsize else s[left: right+1]
        

File: leetcode/76.最小覆盖子串/76.java
class Solution {
    Map<Character, Integer> cnt = new HashMap<Character, Integer>();
    Map<Character, Integer> ori = new HashMap<Character, Integer>();
    public boolean check(){
        Iterator iter = ori.entrySet().iterator();
        while(iter.hasNext()){
            Map.Entry entry =  (Map.Entry) iter.next();
            Integer val = (Integer)entry.getValue();
            Character key = (Character)entry.getKey();
            if(cnt.getOrDefault(key, 0) < val){
                return false;
            }
        } 
        return true;
    }

    public String minWindow(String s, String t) {
        for(int i = 0; i < t.length(); i++){
            char c = t.charAt(i);
            ori.put(c, ori.getOrDefault(c, 0) + 1);
        }
        int i = 0, sub = 0;
        int left = 0, right = 0;
        int res = Integer.MAX_VALUE;
        for(int j = 0; j < s.length(); j++){
            cnt.put(s.charAt(j), cnt.getOrDefault(s.charAt(j), 0) + 1);
            while(check()){
                sub = j - i + 1;
                if(sub < res){
                    left = i;
                    right = j;
                    res = sub;
                }
                if(cnt.containsKey(s.charAt(i))){
                    cnt.put(s.charAt(i), cnt.getOrDefault(s.charAt(i), 0) - 1);
                }
                i++;
            }
        }
        return res == Integer.MAX_VALUE ? "": s.substring(left, right + 1);
    }
}